rollup-node.md

```diff
@@ -0,0 +1,285 @@
+
+[`engine_forkchoiceUpdatedV1`]: https://github.com/ethereum/execution-apis/blob/main/src/engine/specification.md#engine_forkchoiceupdatedv1
+[`engine_forkchoiceUpdatedOPV1`]: TODO
+
+> **TODO LINK** forkchoiceUpdatedOPV1 from spec
+
+- it takes a [`PayloadAttributesOPV1`] object as input instead of
+  [`PayloadAttributesV1`]. The execution engine must include the valid
+  transactions supplied in this object in the block, in the same order as they
+  were supplied, and only those. See the [previous section][payload-attr] for
+  the specification of how the properties must be set.
+
+- we repurpose the [`ForkchoiceStateV1`] structure with the following property
+  semantics:
+  - `headBlockHash`: block hash of the last block of the L2 chain, according to
+    the rollup driver.
+  - `safeBlockHash`: same as `headBlockHash`.
+  - `finalizedBlockHash`: the hash of the block whose number is
+    `number(headBlockHash) - FINALIZATION_DELAY_BLOCKS` if the number of that
+    block is `>= L2_CHAIN_INCEPTION`, 0 otherwise (where
+    `FINALIZATION_DELAY_BLOCKS == 50400` (approximately 7 days worth of L1
+    blocks) and `L2_CHAIN_INCEPTION` is the [L2 chain inception] (the number of
+    the first L1 block for which an L2 block was produced).
+
+[`ForkchoiceStateV1`]: https://github.com/ethereum/execution-apis/blob/main/src/engine/specification.md#ForkchoiceStateV1
+
+> **Note:** the properties of this `ForkchoiceStateV1` can be used to anchor queries
+> to the regular (non-engine-API) JSON-RPC endpoint of the execution engine.
+> [See here for more information.][L2-JSON-RPC-API]
+
+[L2-JSON-RPC-API]: TODO
+
+> **TODO LINK** L2 JSON RPC API (might be the same as [L1's][JSON-RPC-API])
+
+The `payloadID` returned by [`engine_forkchoiceUpdatedOPV1`] can then be passed
+to [`engine_getPayloadV1`] in order to obtain an [`ExecutionPayloadV1`], which
+fully defines a new L2 block.
+
+[`engine_getPayloadV1`]: https://github.com/ethereum/execution-apis/blob/main/src/engine/specification.md#engine_getpayloadv1
+[`ExecutionPayloadV1`]: https://github.com/ethereum/execution-apis/blob/main/src/engine/specification.md#executionpayloadv1
+
+## Handling L1 Re-Orgs
+
+The [previous section on L2 chain derivation][l2-chain-derivation] assumes
+linear progression of the L1 chain. It is also applicable for batch processing,
+meaning that any given point in time, the canonical L2 chain is given by
+processing the whole L1 chain since the [L2 chain inception].
+
+The L1 chain may occasionally re-organize, meaning the head of the L1 chain
+changes to a block that is not the child of the previous head but rather one of
+its "cousins" (i.e. the descendant of an ancestor of the previous head). In
+those case, the rollup driver must:
+
+1. Locate the *common ancestor*, a block that is an ancestor of both the
+   previous and new head.
+2. Isolate the range of L1 blocks \]common ancestor, ..., new head\].
+3. For each such block, call [`engine_forkchoiceUpdatedOPV1`] and
```

Comment:
This is closer to what manual sync would look like, but in some cases it may be better to first try a forkchoice update, and have the execution engine resolve the sync (like it would do in the L1 case).

Reply:
Mmh that is interesting. We should definitely talk about how things work on L1 here.

Reply:
No discussion on this, really agnostic on what it should be. There's still the chance that the upstream spec updates to accomodate our needs.

Reply:
Resolving this, this will be a concern for the execution engine.